### DISCUSSION TITLE
How to prevent XSS attack or token leak using keycloak-js?

### DISCUSSION THREAD

**Discussion by msr-wtag** [keycloak/keycloak#24688] [Q&A]
My application architecture:-
- Frontend authenticates via Keycloak. (so frontend has access to access_token)
- Frontend sends the access_token in authorization header of every api request
- Backend API validates the access_token from keycloak oauth endpoints...
- Thus api is authorized

My question is:-
Since frontend application has access to access_token? What if attackers somehow retrieve the token from client side and get access to all apis???

The architecture is not bulletproof!

**Comment by jonkoops** [comment]
This is always a risk, the best thing to do is not to store your tokens anywhere (such as in local storage). If you are looking to protect yourself from cross-site scripting all conventional wisdom on this topic applies. I'd recommend reading the [OWASP Cross Site Scripting Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html) if you'd like to learn more.
